boostrap.py use curl by default

TApplencourt opened the pull request:
Fixes #61611

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @Mark-Simulacrum (or someone else) soon.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

jyn514 commented:
Please also modify the downloads in `src/bootstrap/native.rs` to use the same fallback.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6248030841?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/fe232454a607a25d7e2bfb58edbe40886501bdb0/checks/6248030841/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
* highest error code: E0787
Found 504 error codes
Found 0 error(s) in error codes
Done!
tidy error: /checkout/src/bootstrap/bootstrap.py:113: line longer than 100 chars
tidy error: /checkout/src/bootstrap/bootstrap.py:126: line longer than 100 chars
tidy error: /checkout/src/bootstrap/bootstrap.py:147: trailing whitespace
some tidy checks failed
Build completed unsuccessfully in 0:00:12
```

</details>

TApplencourt commented:
Thanks for the quick feedback

- Fixed python tidy
- Trying to fix `native.rs`. Sorry for the stupid question, but I was not able to tests my change of `native.rs`. For python I just did `python bootstrap.py`. May I ask how to do the same for `native.rs`? Sorry!

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6248277589?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/14677389a08ecad34aafc696afca694711ee53a3/checks/6248277589/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/src/bootstrap/native.rs at line 310:
     // Try curl, if that fail and we are on windows
     // fallback to PowerShell
     if !builder.try_run(Command::new("curl").args(&[
-            "-#",
-            "-y",
-            "-Y",
-            "-Y",
-            "10", // timeout if speed is < 10 bytes/sec for > 30 seconds
-            "--connect-timeout",
-            "30", // timeout if cannot connect within 30 seconds
-            "--retry",
-            "-Sf",
-            "-o",
-            tempfile,
-            url,
-            url,
-        ])) {
+        "-#",
+        "-y",
+        "30",
+        "-Y",
+        "10", // timeout if speed is < 10 bytes/sec for > 30 seconds
+        "--connect-timeout",
+        "30", // timeout if cannot connect within 30 seconds
+        "--retry",
+        "-Sf",
+        "-o",
+        tempfile,
+        url,
+        url,
+    ])) {
         if builder.build.build.contains("windows-msvc") {
             for _ in 0..3 {
                 if builder.try_run(Command::new("PowerShell.exe").args(&[
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/src/bootstrap/native.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
```

</details>

jyn514 commented:
@TApplencourt you can test the rust changes by setting `[llvm] download-ci-llvm = true` in config.toml, then making sure it successfully downloads LLVM (just `x.py check` should do it, or `x.py build` if you want to be 100% sure rustc_llvm can link to it).

> Sorry for the stupid question, but I was not able to tests my change of native.rs

Not a stupid question! Codebases tend to get more complicated over time :sweat_smile: it's perfectly fine to ask for help figuring out how they work :)

jyn514 commented on src/bootstrap/bootstrap.py:
```diff
@@ -110,29 +110,42 @@ def download(path, url, probably_big, verbose, help_on_error=None):
 
 
 def _download(path, url, probably_big, verbose, exception, help_on_error=None):
+    # Try to use curl (potentialy avalaible on win32
+    #    https://devblogs.microsoft.com/commandline/tar-and-curl-come-to-windows/)
+    # If an error occur:
+    #  - If we are on win32 fallback to powershell
+    #  - Otherwise raise the error if appropriate
```
```suggestion
    # Try to use curl (potentially available on win32
    #    https://devblogs.microsoft.com/commandline/tar-and-curl-come-to-windows/)
    # If an error occurs:
    #  - If we are on win32 fallback to powershell
    #  - Otherwise raise the error if appropriate
```

jyn514 commented on src/bootstrap/bootstrap.py:
```diff
@@ -110,29 +110,42 @@ def download(path, url, probably_big, verbose, help_on_error=None):
+                 "(New-Object System.Net.WebClient).DownloadFile('{}', '{}')".format(url, path)],
+                verbose=verbose,
+                exception=exception)
+        # Aknowledge user will
```
what is this meant to say?

TApplencourt replied:
Bad phrasing indeed. 

I wanted to said that we need to respects the `verbose` or `exception` variables passed as arguments. 
We trapped some errors to perform the fallback mechanism, but some of those errors may need to be raised again.

Does 
```
# Check if the Runtime error of the `run curl` should be raised 
```
sound better?

jyn514 commented on src/bootstrap/native.rs:
```diff
@@ -307,38 +307,39 @@ fn download_component(builder: &Builder<'_>, base: &str, url: &str, dest_path: &
-                ),
-            ])) {
-                return;
+    // Try curl, if that fail and we are on windows
+    // fallback to PowerShell
```
```suggestion
    // Try curl. If that fails and we are on windows, fallback to PowerShell.
```

jyn514 commented on src/bootstrap/native.rs:
```diff
@@ -307,38 +307,39 @@ fn download_component(builder: &Builder<'_>, base: &str, url: &str, dest_path: &
-                return;
+    // Try curl, if that fail and we are on windows
+    // fallback to PowerShell
+    if !builder.try_run(Command::new("curl").args(&[
```
How does this behave on Linux if curl fails? I think currently it will fail the download, but still continue the build, which is not correct.

TApplencourt replied:
Oh good point! It's indeed what the original code is doing. Will fix!

TApplencourt replied:
Added the `check_run` who behave like `try_run` but doesn't terminate

TApplencourt commented on src/bootstrap/native.rs:
```diff
@@ -306,39 +306,40 @@ fn download_component(builder: &Builder<'_>, base: &str, url: &str, dest_path: &
-            tempfile,
-            url,
-        ]));
+        std::process::exit(1);
```
I think that fix a bug where before, if the 3 try to PowerSheel failed, we didn't terminate

Mark-Simulacrum commented on src/bootstrap/bootstrap.py:
```diff
@@ -110,29 +110,42 @@ def download(path, url, probably_big, verbose, help_on_error=None):
+                exception=exception)
+        # Check if the RuntimeError raised by run(curl) should be silenced
+        elif verbose or exception:
+            raise RuntimeError
```
It seems like we should be trying to re-raise the original exception -- this looks like it'll probably create a new exception instead?

TApplencourt replied:
TLDR: Good catch! I will remove the `RuntimeError`.

Long note:
In python3  the traceback is stored in the exception:
```
$cat test.py
def foo():
	raise RuntimeError("foo error")
try:
	foo()
except RuntimeError:
	raise RuntimeError
$python3 test.py
Traceback (most recent call last):
  File "test.py", line 5, in <module>
    foo()
  File "test.py", line 2, in foo
    raise RuntimeError("foo error")
RuntimeError: foo error

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test.py", line 7, in <module>
    raise RuntimeError
RuntimeError
```

But the traceback is ugly. With just `raise` everything is nice and tidy
 ```
$cat test.py
def foo():
	raise RuntimeError("foo error")
try:
	foo()
except RuntimeError:
	raise
$python3 test.py
Traceback (most recent call last):
  File "test.py", line 5, in <module>
    foo()
  File "test.py", line 2, in foo
    raise RuntimeError("foo error")
RuntimeError: foo error
(base) razoa@hagarth:~/ru
```

Thanks!

Mark-Simulacrum reviewed (commented):
Please also keep commits squashed. Thanks!

TApplencourt commented:
I think I addressed all the comments. Thanks a lot for the feedback! 

I did some checks on my Linux machine; if `curl` is not present, we will check if we are on windows and if so try the `PowerShell` way.  

I'm not sure about the naming of `check_run` (the `try_run` who never aborts).  I can change it to any other more descriptive name if needed.

Thanks again for your time and your guidance!

TApplencourt commented:
@jyn514, @Mark-Simulacrum  If you have time, I think this is ready to review! 
 Sorry, if I'm too pushy 🙏🏽

Mark-Simulacrum commented:
@bors r+

bors commented:
:pushpin: Commit ad7dbe18da68aff54de4315274a7ca8442d08ed8 has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum ad7dbe18da68aff54de4315274a7ca8442d08ed8 -->
<!-- homu: {"type":"Approved","sha":"ad7dbe18da68aff54de4315274a7ca8442d08ed8","approver":"Mark-Simulacrum"} -->

bors commented:
:hourglass: Testing commit ad7dbe18da68aff54de4315274a7ca8442d08ed8 with merge 0be876832331883a0e9057495dc00d855d53097a...
<!-- homu: {"type":"BuildStarted","head_sha":"ad7dbe18da68aff54de4315274a7ca8442d08ed8","merge_sha":"0be876832331883a0e9057495dc00d855d53097a"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/6439571167?check_suite_focus=true)
Approved by: Mark-Simulacrum
Pushing 0be876832331883a0e9057495dc00d855d53097a to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Mark-Simulacrum","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/6439571167?check_suite_focus=true"},"merge_sha":"0be876832331883a0e9057495dc00d855d53097a"} -->

rust-timer commented:
Finished benchmarking commit (0be876832331883a0e9057495dc00d855d53097a): [comparison url](https://perf.rust-lang.org/compare.html?start=7fb2d0ecef572120c4f2fc9a12a23f9bc9ace3d9&end=0be876832331883a0e9057495dc00d855d53097a).

**Summary**: This benchmark run did not return any relevant results.

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).



@rustbot label: -perf-regression

TApplencourt commented:
Thanks a lot y'all for the guidance on this one!

jyn514 commented:
Thank you for working on it!
